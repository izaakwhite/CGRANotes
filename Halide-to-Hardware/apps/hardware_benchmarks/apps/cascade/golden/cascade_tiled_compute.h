#pragma once
#include "hw_classes.h"
#include "clockwork_standard_compute_units.h"


//store is: hw_input_global_wrapper.stencil(hw_input_global_wrapper_s0_x, hw_input_global_wrapper_s0_y) = hw_input.stencil(hw_input_global_wrapper_s0_x, hw_input_global_wrapper_s0_y)
hw_uint<16> hcompute_hw_input_global_wrapper_stencil(hw_uint<16>& hw_input_stencil) {
  uint16_t _hw_input_stencil_1 = (uint16_t) hw_input_stencil.extract<0, 15>();

  return _hw_input_stencil_1;
}

//store is: conv1.stencil(conv1_s0_x, conv1_s0_y) = (uint16)0
hw_uint<16> hcompute_conv1_stencil() {
  uint16_t _284 = (uint16_t)(0);
  return _284;
}

//store is: conv1.stencil(conv1_s1_x, conv1_s1_y) = (hw_input_global_wrapper.stencil(conv1_s1_x, conv1_s1_y) + (conv1.stencil(conv1_s1_x, conv1_s1_y) + ((hw_input_global_wrapper.stencil(conv1_s1_x, (conv1_s1_y + 1))*(uint16)2) + (hw_input_global_wrapper.stencil(conv1_s1_x, (conv1_s1_y + 2)) + ((hw_input_global_wrapper.stencil((conv1_s1_x + 1), conv1_s1_y)*(uint16)2) + ((hw_input_global_wrapper.stencil((conv1_s1_x + 1), (conv1_s1_y + 1))*(uint16)4) + ((hw_input_global_wrapper.stencil((conv1_s1_x + 1), (conv1_s1_y + 2))*(uint16)2) + (hw_input_global_wrapper.stencil((conv1_s1_x + 2), conv1_s1_y) + (hw_input_global_wrapper.stencil((conv1_s1_x + 2), (conv1_s1_y + 2)) + (hw_input_global_wrapper.stencil((conv1_s1_x + 2), (conv1_s1_y + 1))*(uint16)2))))))))))
hw_uint<16> hcompute_conv1_stencil_1(hw_uint<16>& conv1_stencil, hw_uint<144>& hw_input_global_wrapper_stencil) {
  uint16_t _conv1_stencil_1 = (uint16_t) conv1_stencil.extract<0, 15>();

  uint16_t _hw_input_global_wrapper_stencil_1 = (uint16_t) hw_input_global_wrapper_stencil.extract<0, 15>();
  uint16_t _hw_input_global_wrapper_stencil_2 = (uint16_t) hw_input_global_wrapper_stencil.extract<16, 31>();
  uint16_t _hw_input_global_wrapper_stencil_3 = (uint16_t) hw_input_global_wrapper_stencil.extract<32, 47>();
  uint16_t _hw_input_global_wrapper_stencil_4 = (uint16_t) hw_input_global_wrapper_stencil.extract<48, 63>();
  uint16_t _hw_input_global_wrapper_stencil_5 = (uint16_t) hw_input_global_wrapper_stencil.extract<64, 79>();
  uint16_t _hw_input_global_wrapper_stencil_6 = (uint16_t) hw_input_global_wrapper_stencil.extract<80, 95>();
  uint16_t _hw_input_global_wrapper_stencil_7 = (uint16_t) hw_input_global_wrapper_stencil.extract<96, 111>();
  uint16_t _hw_input_global_wrapper_stencil_8 = (uint16_t) hw_input_global_wrapper_stencil.extract<112, 127>();
  uint16_t _hw_input_global_wrapper_stencil_9 = (uint16_t) hw_input_global_wrapper_stencil.extract<128, 143>();

  uint16_t _287 = (uint16_t)(2);
  uint16_t _288 = _hw_input_global_wrapper_stencil_2 * _287;
  uint16_t _289 = _hw_input_global_wrapper_stencil_4 * _287;
  uint16_t _290 = (uint16_t)(4);
  uint16_t _291 = _hw_input_global_wrapper_stencil_5 * _290;
  uint16_t _292 = _hw_input_global_wrapper_stencil_6 * _287;
  uint16_t _293 = _hw_input_global_wrapper_stencil_9 * _287;
  uint16_t _294 = _hw_input_global_wrapper_stencil_8 + _293;
  uint16_t _295 = _hw_input_global_wrapper_stencil_7 + _294;
  uint16_t _296 = _292 + _295;
  uint16_t _297 = _291 + _296;
  uint16_t _298 = _289 + _297;
  uint16_t _299 = _hw_input_global_wrapper_stencil_3 + _298;
  uint16_t _300 = _288 + _299;
  uint16_t _301 = _conv1_stencil_1 + _300;
  uint16_t _302 = _hw_input_global_wrapper_stencil_1 + _301;
  return _302;
}

//store is: conv2.stencil(conv2_s0_x, conv2_s0_y) = (uint16)0
hw_uint<16> hcompute_conv2_stencil() {
  uint16_t _349 = (uint16_t)(0);
  return _349;
}

//store is: conv2.stencil(conv2_s1_x, conv2_s1_y) = (conv1.stencil(conv2_s1_x, conv2_s1_y) + (conv2.stencil(conv2_s1_x, conv2_s1_y) + ((conv1.stencil((conv2_s1_x + 1), conv2_s1_y)*(uint16)2) + (conv1.stencil((conv2_s1_x + 2), conv2_s1_y) + ((conv1.stencil(conv2_s1_x, (conv2_s1_y + 1))*(uint16)2) + ((conv1.stencil((conv2_s1_x + 1), (conv2_s1_y + 1))*(uint16)4) + ((conv1.stencil((conv2_s1_x + 2), (conv2_s1_y + 1))*(uint16)2) + (conv1.stencil(conv2_s1_x, (conv2_s1_y + 2)) + (conv1.stencil((conv2_s1_x + 2), (conv2_s1_y + 2)) + (conv1.stencil((conv2_s1_x + 1), (conv2_s1_y + 2))*(uint16)2))))))))))
hw_uint<16> hcompute_conv2_stencil_1(hw_uint<144>& conv1_stencil, hw_uint<16>& conv2_stencil) {
  uint16_t _conv1_stencil_10 = (uint16_t) conv1_stencil.extract<0, 15>();
  uint16_t _conv1_stencil_2 = (uint16_t) conv1_stencil.extract<16, 31>();
  uint16_t _conv1_stencil_3 = (uint16_t) conv1_stencil.extract<32, 47>();
  uint16_t _conv1_stencil_4 = (uint16_t) conv1_stencil.extract<48, 63>();
  uint16_t _conv1_stencil_5 = (uint16_t) conv1_stencil.extract<64, 79>();
  uint16_t _conv1_stencil_6 = (uint16_t) conv1_stencil.extract<80, 95>();
  uint16_t _conv1_stencil_7 = (uint16_t) conv1_stencil.extract<96, 111>();
  uint16_t _conv1_stencil_8 = (uint16_t) conv1_stencil.extract<112, 127>();
  uint16_t _conv1_stencil_9 = (uint16_t) conv1_stencil.extract<128, 143>();

  uint16_t _conv2_stencil_1 = (uint16_t) conv2_stencil.extract<0, 15>();

  uint16_t _352 = (uint16_t)(2);
  uint16_t _353 = _conv1_stencil_3 * _352;
  uint16_t _354 = _conv1_stencil_5 * _352;
  uint16_t _355 = (uint16_t)(4);
  uint16_t _356 = _conv1_stencil_6 * _355;
  uint16_t _357 = _conv1_stencil_7 * _352;
  uint16_t _358 = _conv1_stencil_10 * _352;
  uint16_t _359 = _conv1_stencil_9 + _358;
  uint16_t _360 = _conv1_stencil_8 + _359;
  uint16_t _361 = _357 + _360;
  uint16_t _362 = _356 + _361;
  uint16_t _363 = _354 + _362;
  uint16_t _364 = _conv1_stencil_4 + _363;
  uint16_t _365 = _353 + _364;
  uint16_t _366 = _conv2_stencil_1 + _365;
  uint16_t _367 = _conv1_stencil_2 + _366;
  return _367;
}

//store is: hw_output.stencil(hw_output_s0_x_xi, hw_output_s0_y_yi) = conv2.stencil(hw_output_s0_x_xi, hw_output_s0_y_yi)
hw_uint<16> hcompute_hw_output_stencil(hw_uint<16>& conv2_stencil) {
  uint16_t _conv2_stencil_2 = (uint16_t) conv2_stencil.extract<0, 15>();

  return _conv2_stencil_2;
}

