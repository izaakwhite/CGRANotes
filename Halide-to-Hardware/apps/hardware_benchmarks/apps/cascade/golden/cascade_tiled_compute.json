{
"namespaces":{
  "global":{
    "modules":{
      "hcompute_conv1_stencil":{
        "type":["Record",[
          ["out_conv1_stencil",["Array",16,"Bit"]]
        ]],
        "instances":{
          "const_p0__285":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0000"]}
          }
        },
        "connections":[
          ["self.out_conv1_stencil","const_p0__285.out"]
        ]
      },
      "hcompute_conv1_stencil_1":{
        "type":["Record",[
          ["out_conv1_stencil",["Array",16,"Bit"]],
          ["in0_conv1_stencil",["Array",1,["Array",16,"BitIn"]]],
          ["in1_hw_input_global_wrapper_stencil",["Array",9,["Array",16,"BitIn"]]]
        ]],
        "instances":{
          "add_304_315_316":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_305_313_314":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_307_312_313":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_308_311_312":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv1_stencil_1_316_317":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_hw_input_global_wrapper_stencil_1_317_318":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_hw_input_global_wrapper_stencil_3_314_315":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_hw_input_global_wrapper_stencil_7_310_311":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_hw_input_global_wrapper_stencil_8_309_310":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "const_p2__303":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__303$1":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__303$2":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__303$3":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p4__306":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0004"]}
          },
          "mul_hw_input_global_wrapper_stencil_2_303_304":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_hw_input_global_wrapper_stencil_4_303_305":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_hw_input_global_wrapper_stencil_5_306_307":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_hw_input_global_wrapper_stencil_6_303_308":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_hw_input_global_wrapper_stencil_9_303_309":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          }
        },
        "connections":[
          ["mul_hw_input_global_wrapper_stencil_2_303_304.out","add_304_315_316.in0"],
          ["add_hw_input_global_wrapper_stencil_3_314_315.out","add_304_315_316.in1"],
          ["add_conv1_stencil_1_316_317.in1","add_304_315_316.out"],
          ["mul_hw_input_global_wrapper_stencil_4_303_305.out","add_305_313_314.in0"],
          ["add_307_312_313.out","add_305_313_314.in1"],
          ["add_hw_input_global_wrapper_stencil_3_314_315.in1","add_305_313_314.out"],
          ["mul_hw_input_global_wrapper_stencil_5_306_307.out","add_307_312_313.in0"],
          ["add_308_311_312.out","add_307_312_313.in1"],
          ["mul_hw_input_global_wrapper_stencil_6_303_308.out","add_308_311_312.in0"],
          ["add_hw_input_global_wrapper_stencil_7_310_311.out","add_308_311_312.in1"],
          ["self.in0_conv1_stencil.0","add_conv1_stencil_1_316_317.in0"],
          ["add_hw_input_global_wrapper_stencil_1_317_318.in1","add_conv1_stencil_1_316_317.out"],
          ["self.in1_hw_input_global_wrapper_stencil.0","add_hw_input_global_wrapper_stencil_1_317_318.in0"],
          ["self.out_conv1_stencil","add_hw_input_global_wrapper_stencil_1_317_318.out"],
          ["self.in1_hw_input_global_wrapper_stencil.2","add_hw_input_global_wrapper_stencil_3_314_315.in0"],
          ["self.in1_hw_input_global_wrapper_stencil.6","add_hw_input_global_wrapper_stencil_7_310_311.in0"],
          ["add_hw_input_global_wrapper_stencil_8_309_310.out","add_hw_input_global_wrapper_stencil_7_310_311.in1"],
          ["self.in1_hw_input_global_wrapper_stencil.7","add_hw_input_global_wrapper_stencil_8_309_310.in0"],
          ["mul_hw_input_global_wrapper_stencil_9_303_309.out","add_hw_input_global_wrapper_stencil_8_309_310.in1"],
          ["mul_hw_input_global_wrapper_stencil_4_303_305.in1","const_p2__303$1.out"],
          ["mul_hw_input_global_wrapper_stencil_6_303_308.in1","const_p2__303$2.out"],
          ["mul_hw_input_global_wrapper_stencil_9_303_309.in1","const_p2__303$3.out"],
          ["mul_hw_input_global_wrapper_stencil_2_303_304.in1","const_p2__303.out"],
          ["mul_hw_input_global_wrapper_stencil_5_306_307.in1","const_p4__306.out"],
          ["self.in1_hw_input_global_wrapper_stencil.1","mul_hw_input_global_wrapper_stencil_2_303_304.in0"],
          ["self.in1_hw_input_global_wrapper_stencil.3","mul_hw_input_global_wrapper_stencil_4_303_305.in0"],
          ["self.in1_hw_input_global_wrapper_stencil.4","mul_hw_input_global_wrapper_stencil_5_306_307.in0"],
          ["self.in1_hw_input_global_wrapper_stencil.5","mul_hw_input_global_wrapper_stencil_6_303_308.in0"],
          ["self.in1_hw_input_global_wrapper_stencil.8","mul_hw_input_global_wrapper_stencil_9_303_309.in0"]
        ]
      },
      "hcompute_conv2_stencil":{
        "type":["Record",[
          ["out_conv2_stencil",["Array",16,"Bit"]]
        ]],
        "instances":{
          "const_p0__350":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0000"]}
          }
        },
        "connections":[
          ["self.out_conv2_stencil","const_p0__350.out"]
        ]
      },
      "hcompute_conv2_stencil_1":{
        "type":["Record",[
          ["out_conv2_stencil",["Array",16,"Bit"]],
          ["in0_conv1_stencil",["Array",9,["Array",16,"BitIn"]]],
          ["in1_conv2_stencil",["Array",1,["Array",16,"BitIn"]]]
        ]],
        "instances":{
          "add_369_380_381":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_370_378_379":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_372_377_378":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_373_376_377":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv1_stencil_2_382_383":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv1_stencil_4_379_380":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv1_stencil_8_375_376":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv1_stencil_9_374_375":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "add_conv2_stencil_1_381_382":{
            "genref":"coreir.add",
            "genargs":{"width":["Int",16]}
          },
          "const_p2__368":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__368$1":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__368$2":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p2__368$3":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0002"]}
          },
          "const_p4__371":{
            "genref":"coreir.const",
            "genargs":{"width":["Int",16]},
            "modargs":{"value":[["BitVector",16],"16'h0004"]}
          },
          "mul_conv1_stencil_10_368_374":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_conv1_stencil_3_368_369":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_conv1_stencil_5_368_370":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_conv1_stencil_6_371_372":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          },
          "mul_conv1_stencil_7_368_373":{
            "genref":"coreir.mul",
            "genargs":{"width":["Int",16]}
          }
        },
        "connections":[
          ["mul_conv1_stencil_3_368_369.out","add_369_380_381.in0"],
          ["add_conv1_stencil_4_379_380.out","add_369_380_381.in1"],
          ["add_conv2_stencil_1_381_382.in1","add_369_380_381.out"],
          ["mul_conv1_stencil_5_368_370.out","add_370_378_379.in0"],
          ["add_372_377_378.out","add_370_378_379.in1"],
          ["add_conv1_stencil_4_379_380.in1","add_370_378_379.out"],
          ["mul_conv1_stencil_6_371_372.out","add_372_377_378.in0"],
          ["add_373_376_377.out","add_372_377_378.in1"],
          ["mul_conv1_stencil_7_368_373.out","add_373_376_377.in0"],
          ["add_conv1_stencil_8_375_376.out","add_373_376_377.in1"],
          ["self.in0_conv1_stencil.1","add_conv1_stencil_2_382_383.in0"],
          ["add_conv2_stencil_1_381_382.out","add_conv1_stencil_2_382_383.in1"],
          ["self.out_conv2_stencil","add_conv1_stencil_2_382_383.out"],
          ["self.in0_conv1_stencil.3","add_conv1_stencil_4_379_380.in0"],
          ["self.in0_conv1_stencil.7","add_conv1_stencil_8_375_376.in0"],
          ["add_conv1_stencil_9_374_375.out","add_conv1_stencil_8_375_376.in1"],
          ["self.in0_conv1_stencil.8","add_conv1_stencil_9_374_375.in0"],
          ["mul_conv1_stencil_10_368_374.out","add_conv1_stencil_9_374_375.in1"],
          ["self.in1_conv2_stencil.0","add_conv2_stencil_1_381_382.in0"],
          ["mul_conv1_stencil_5_368_370.in1","const_p2__368$1.out"],
          ["mul_conv1_stencil_7_368_373.in1","const_p2__368$2.out"],
          ["mul_conv1_stencil_10_368_374.in1","const_p2__368$3.out"],
          ["mul_conv1_stencil_3_368_369.in1","const_p2__368.out"],
          ["mul_conv1_stencil_6_371_372.in1","const_p4__371.out"],
          ["self.in0_conv1_stencil.0","mul_conv1_stencil_10_368_374.in0"],
          ["self.in0_conv1_stencil.2","mul_conv1_stencil_3_368_369.in0"],
          ["self.in0_conv1_stencil.4","mul_conv1_stencil_5_368_370.in0"],
          ["self.in0_conv1_stencil.5","mul_conv1_stencil_6_371_372.in0"],
          ["self.in0_conv1_stencil.6","mul_conv1_stencil_7_368_373.in0"]
        ]
      },
      "hcompute_hw_input_global_wrapper_stencil":{
        "type":["Record",[
          ["out_hw_input_global_wrapper_stencil",["Array",16,"Bit"]],
          ["in0_hw_input_stencil",["Array",1,["Array",16,"BitIn"]]]
        ]],
        "connections":[
          ["self.out_hw_input_global_wrapper_stencil","self.in0_hw_input_stencil.0"]
        ]
      },
      "hcompute_hw_output_stencil":{
        "type":["Record",[
          ["out_hw_output_stencil",["Array",16,"Bit"]],
          ["in0_conv2_stencil",["Array",1,["Array",16,"BitIn"]]]
        ]],
        "connections":[
          ["self.out_hw_output_stencil","self.in0_conv2_stencil.0"]
        ]
      }
    }
  }
}
}
